cmd/notification.go

```diff
@@ -0,0 +1,634 @@
+	sys.Lock()
+	defer sys.Unlock()
+
+	delete(sys.bucketRulesMap, bucketName)
+
+	for targetID := range sys.bucketRemoteTargetRulesMap[bucketName] {
+		sys.targetList.Remove(targetID)
+		delete(sys.bucketRemoteTargetRulesMap[bucketName], targetID)
+	}
+
+	delete(sys.bucketRemoteTargetRulesMap, bucketName)
+}
+
+// RemoveAllRemoteTargets - closes and removes all HTTP/PeerRPC client targets.
+func (sys *NotificationSys) RemoveAllRemoteTargets() {
+	for _, targetMap := range sys.bucketRemoteTargetRulesMap {
+		for targetID := range targetMap {
+			sys.targetList.Remove(targetID)
+		}
+	}
+}
+
+// RemoveRemoteTarget - closes and removes target by target ID.
+func (sys *NotificationSys) RemoveRemoteTarget(bucketName string, targetID event.TargetID) {
+	for id, err := range sys.targetList.Remove(targetID) {
+		errorIf(err, "unable to close target ID %v", id)
+	}
+
+	sys.Lock()
+	defer sys.Unlock()
+
+	if _, ok := sys.bucketRemoteTargetRulesMap[bucketName]; ok {
+		delete(sys.bucketRemoteTargetRulesMap[bucketName], targetID)
+		if len(sys.bucketRemoteTargetRulesMap[bucketName]) == 0 {
+			delete(sys.bucketRemoteTargetRulesMap, bucketName)
+		}
+	}
+}
+
+func (sys *NotificationSys) send(bucketName string, eventData event.Event, targetIDs ...event.TargetID) map[event.TargetID]error {
+	errMap := sys.targetList.Send(eventData, targetIDs...)
+	for targetID := range errMap {
+		if sys.RemoteTargetExist(bucketName, targetID) {
+			sys.RemoveRemoteTarget(bucketName, targetID)
+		}
+	}
+
+	return errMap
+}
+
+// Send - sends event data to all matching targets.
+func (sys *NotificationSys) Send(args eventArgs) map[event.TargetID]error {
+	sys.RLock()
+	targetIDSet := sys.bucketRulesMap[args.BucketName].Match(args.EventName, args.Object.Name)
+	sys.RUnlock()
+	if len(targetIDSet) == 0 {
+		return nil
+	}
+
+	targetIDs := targetIDSet.ToSlice()
+	return sys.send(args.BucketName, args.ToEvent(), targetIDs...)
+}
+
+// NewNotificationSys - creates new notification system object.
+func NewNotificationSys(config *serverConfig, endpoints EndpointList) (*NotificationSys, error) {
+	targetList, err := getNotificationTargets(config)
+	if err != nil {
+		return nil, err
+	}
+
+	peerRPCClientMap := makeRemoteRPCClients(endpoints)
+
+	// bucketRulesMap/bucketRemoteTargetRulesMap are initialized by NotificationSys.Init()
+	return &NotificationSys{
+		targetList:                 targetList,
+		bucketRulesMap:             make(map[string]event.RulesMap),
+		bucketRemoteTargetRulesMap: make(map[string]map[event.TargetID]event.RulesMap),
+		peerRPCClientMap:           peerRPCClientMap,
+	}, nil
+}
+
+type eventArgs struct {
+	EventName  event.Name
+	BucketName string
+	Object     ObjectInfo
+	ReqParams  map[string]string
+	Host       string
+	Port       string
+	UserAgent  string
+}
+
+// ToEvent - converts to notification event.
+func (args eventArgs) ToEvent() event.Event {
+	getOriginEndpoint := func() string {
+		host := globalMinioHost
+		if host == "" {
```

Comment:
Looks like this won't be empty anymore with the change in endpoint.go?

Reply:
Not sure how `endpoint.go` fixes the issue.  `globalMinioHost` may be empty in FS/XL.